porcupine/actions.py

```diff
@@ -0,0 +1,140 @@
+from dataclasses import dataclass
+from functools import partial
+from pathlib import Path
+from typing import Callable
+
+from porcupine.tabs import FileTab
+
+action_availability_callback = Callable[[], bool]
+
+
+@dataclass(kw_only=True, slots=True, frozen=True)
+class Action:
+    """Action that requires no context in the callback"""
+
+    name: str
+    description: str
+    callback: Callable[..., None]
+    availability_callbacks: list[action_availability_callback] | None = None
```

Comment:
I would prefer just `Callable[[], bool]`, not a list. There's also no need for `None` because you can use `lambda: True`. This is how `porcupine/plugins/filemanager.py` works now. It basically defines a bunch of `PathAction`s.

```python
commands = [
    # Doing something to an entire project is more difficult than you would think.
    # For example, if the project is renamed, venv locations don't update.
    # TODO: update venv locations when the venv is renamed
    Command("New file here", "<<FileManager:New file>>", (lambda p: True), new_file_here),
    Command(
        "New directory here", "<<FileManager:New directory>>", (lambda p: True), new_directory_here
    ),
    Command("Cut", "<<Cut>>", (lambda p: not p.is_dir()), cut),
    Command("Copy", "<<Copy>>", (lambda p: not p.is_dir()), copy),
    Command("Paste", "<<Paste>>", can_paste, paste),
    Command("Rename", "<<FileManager:Rename>>", is_NOT_project_root, rename),
    Command(f"Move to {trash_name}", "<<FileManager:Trash>>", is_NOT_project_root, trash),
    Command("Delete", "<<FileManager:Delete>>", is_NOT_project_root, delete),
    Command("Open in file manager", None, (lambda p: p.is_dir()), open_in_file_manager),
    Command("Open in terminal", None, (lambda p: p.is_dir()), open_in_terminal),
    Command("Copy full path to clipboard", None, (lambda p: True), copy_full_path_to_clipboard),
]
```

Reply:
for the `availability_callbacks`?

I think that the lists will help a lot with DX because we can create helper utilities to implement common uses, and you can compose them in the list.

Consider the different things that someone might need to check to see if a action is available:
- filetype of tab
- filetype of file
- name of file
- if a Path is the "project root"
- if a Path is `.gitignore`'d
- contents of the FileTab selection
- contents of the FileTab `insert` line
- contents of the clipboard
- whether a FileTab has unsaved changes
- contents of the Filetab undo stack
- if internet is available
- various config options
- if plugins are available

I'm sure there are more.

The thought is instead of each action plugin having to create their own availability callbacks, we could let people "compose" them in a list from the helpers we created. Obviously if something is needed that doesn't exist the plugin dev could use that too.

```python
register_path_action(
    name="Update requirements.txt",
    description="Generate requirements.txt from `pyproject.toml`",
    callback=...,
    availability_callbacks=[plugin_availability("requirements_update"), internet_availability(), file_extension_availability(["txt"]), file_name_availability(r"requirements*.txt"), filetype_availability("Python Requirements") ],
)
```

You might think that the above is contrived, but I don't think so. The idea also is that the availability checks will execute from the lowest index in the list up, so you would want to put the cheap and broad checks at the beginning to make sure it's fast.

Reply:
It sounds to me that we're reinventing callback functions. With `def requirements_update_is_available() -> bool`, it is very obvious how the thing works, and you can still reuse existing things easily.

Reply:
Can you expand on that? Are you saying instead of:
```python
    availability_callbacks=[plugin_availability("requirements_update"), internet_availability(), file_extension_availability(["txt"]), file_name_availability(r"requirements*.txt"), filetype_availability("Python Requirements") ],
```

You would suggest:
```python
def requirements_update_is_available() -> bool:
    for callback in [plugin_availability("requirements_update"), internet_availability(), file_extension_availability(["txt"]), file_name_availability(r"requirements*.txt"), filetype_availability("Python Requirements") ]:
        available = callback()
        if not available:
            return False

    return True

...
    availability_callback=requirements_update_is_available
```
?

Obviously that does work, but I'm not really sure I'd prefer it..

Reply:
More like:
```python3
def requirements_update_is_available() -> bool:
    return (
        plugin_is_enabled("requirements_update")
        and connected_to_internet()
        and get_file_extension(get_current_tab().path) == "txt"
        and glob.match(something r"requirements*.txt")
        and get_current_filetype() == "Python Requirements"
    )

...
    availability_callback=requirements_update_is_available
```

Or if you prefer:

```python3
...
    availability_callback=lambda: (
        plugin_is_enabled("requirements_update")
        and connected_to_internet()
        and get_file_extension(get_current_tab().path) == "txt"
        and glob.match(something r"requirements*.txt")
        and get_current_filetype() == "Python Requirements"
    )
```

One more advantage is the explicit `and` keyword. With a list, you don't know whether it's `or`ing or `and`ing the individual conditions. IMO both can be what the user wants, so it's better to be explicit.

Reply:
Okay, that makes sense, and it does cover the case I was trying to cover of executing from right to left only until the first falsey value:

```python
def false():
    print("false")
    return False


def true():
    print("true")
    return True


print(false() and true())
```

returns:
```
false
False
```

So if you have expensive availability callbacks, you can put them at the end and they won't run unless everything else passes :)

Reply:
Yes, this pattern is more common in some other languages (e.g. C) but also works in python :)

Reply:
Wait actually this doesn't quite work as you have conceived it, I think?
```python
...
    availability_callback=lambda: (
        plugin_is_enabled("requirements_update")
        and connected_to_internet()
        and get_file_extension(get_current_tab().path) == "txt"
        and glob.match(something r"requirements*.txt")
        and get_current_filetype() == "Python Requirements"
    )
```

what context do these execute? they execute at import time, right? maybe I'm confused... but taking my example of `filetype_availability("Python Requirements")`, that **returns a `Callable`**, not a bool, so wouldn't `and` not work?

In fact, I'm realizing my example was wrong based on that, it would need to be:
```python
def requirements_update_is_available() -> bool:
    for callback in [plugin_availability("requirements_update"), internet_availability(), file_extension_availability(["txt"]), file_name_availability(r"requirements*.txt"), filetype_availability("Python Requirements") ]:
        available = callback(context=context) # this is the line that changes 
        if not available:
            return False

    return True

...
    availability_callback=requirements_update_is_available
```

aka the availability callbacks get injected with the same context that the actual action callback does (FileTab, Path, or nothing as of now).

Does this work the way you have concieved?

Reply:
Because it is inside a lambda, it executes when the callback is called, i.e. when the availability needs to be determined. For actions that take context (so anything except `BareAction`), it would be `lambda context: (... and ... and ...)`. That's why the list in `filemanager.py` has `lambda p: ...` conditions.

Reply:
Ah, now I see your confusion:

>filetype_availability("Python Requirements"), that returns a Callable, not a bool, so wouldn't and not work?

If it returns a callback, it would look like `... and filetype_availability("Python Requirements")(path) and ...`. Not very nice. I would just rewrite the availability function so that I can write `... and get_filetype_name_for_path(path) == "Python Requirements" and ...`

Reply:
so would each of them need to be partials?

```python
...
    availability_callback=lambda tab: (
        partial(plugin_is_enabled("requirements_update"), tab)
        and connected_to_internet()
        and get_file_extension(tab.path) == "txt"
        and glob.match(tab.path.filename() r"requirements*.txt")
        and partial(get_current_filetype(), tab) == "Python Requirements"
    )
```

?

One of us is missing something (probably me)

Reply:
None of them would be partials, because `and`ing `Callable`s doesn't work. Instead we invoke all `Callable`s inside the lambda until we get `bool`s for `and`ing.

For example, consider `connected_to_internet()`. It is a plain old function that returns a boolean: `def connected_to_internet() -> bool`. You use it in the `and` chain just like anything else.

Because this is wrapped in the `lambda`, it doesn't get evaluated right away, and if the user connects to the internet after starting Porcupine, the action will become available as expected.

Reply:
Okay, so I guess it would actually be like this:
```python
...
    availability_callback=lambda tab: (
        plugin_is_enabled("requirements_update")
        and connected_to_internet()
        and get_file_extension(tab.path) == "txt"
        and glob.match(tab.path.filename(), r"requirements*.txt")
        and get_current_filetype(tab) == "Python Requirements"
    )
```

So no partials needed?

Reply:
Yes, that's how it would work :)

Reply:
Okay thanks! I will write a couple of these proposed availability functions and see if it all starts to come together, but I think we are on the same page now :)